### PR TITLE
[feat] Added mix augmentations in containers

### PR DESCRIPTION
**Pull request by shijianjian**
This PR enabled mix augmentations in containers.

In order to process labels correctly, I forced to apply only one mix augmentation at the time.

Sorry for the delay that I have refactored many times to get a decent API. The whole general structure is more or less fixed, and I will update the internal mess shortly.

\cc @LinShiqi047

### CODE REVIEW

File: kornia/augmentation/base.py [hunk 1]
```diff
@@ -112,9 +112,7 @@ def forward_parameters(self, batch_shape):
         _params['batch_prob'] = to_apply
         return _params
 
-    def apply_func(
-        self, input: torch.Tensor, params: Dict[str, torch.Tensor]
-    ) -> TensorWithTransMat:
+    def apply_func(self, input: torch.Tensor, params: Dict[str, torch.Tensor]) -> TensorWithTransMat:
```
Comment (edgarriba) [inline kornia/augmentation/base.py:115]: `TensorWithTransMat` -> `TensorWithTransformMat`

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -1911,14 +1911,15 @@ def __init__(
...
+        self.max_val = max_val
 
     def __repr__(self) -> str:
         return self.__class__.__name__ + f"({super().__repr__()})"
 
     def apply_transform(
         self, input: torch.Tensor, params: Dict[str, torch.Tensor], transform: Optional[torch.Tensor] = None
     ) -> torch.Tensor:
-        return invert(input)
+        return invert(input, torch.as_tensor(self.max_val, device=input.device, dtype=input.dtype))
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:1922]: if we assume that `max_val` can be a float we should update typing too.

File: kornia/augmentation/container/image.py [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
-from kornia.augmentation.base import _AugmentationBase
+from kornia.augmentation.base import _AugmentationBase, MixAugmentationBase, TensorWithTransMat
 
 __all__ = ["ImageSequential"]
 
 
 class ParamItem(NamedTuple):
     name: str
-    data: Union[dict, list]
+    data: Optional[Union[dict, list]]
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:16]: when data can be None ?

File: kornia/augmentation/container/image.py [hunk 4]
```diff
@@ -35,29 +37,35 @@ class ImageSequential(nn.Sequential):
             If False, the whole list of args will be processed as a sequence in original order.
 
     Returns:
-        the tensor (, and the transformation matrix) has been sequentially modified by the args.
+        TensorWithTransMat: the tensor (, and the transformation matrix)
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:40]:
```suggestion
        the transformed image tensor (, and the transformation matrix)
```

Comment (edgarriba) [reply kornia/augmentation/container/image.py:40]: types no needed to add anymore

File: kornia/augmentation/container/image.py [hunk 5]
```diff
@@ -35,29 +37,35 @@ class ImageSequential(nn.Sequential):
...
+            has been sequentially modified by the args.
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:41]: don't understand this sentence. The return of this container is the transformed image and the transform matrix, right ?

File: kornia/augmentation/container/image.py [hunk 6]
```diff
@@ -71,24 +79,19 @@ def __init__(
...
     ) -> None:
-        self.same_on_batch = same_on_batch
-        self.return_transform = return_transform
-        self.keepdim = keepdim
+        self._same_on_batch = same_on_batch
+        self._return_transform = return_transform
+        self._keepdim = keepdim
         # To name the modules properly
         _args = OrderedDict()
         for idx, arg in enumerate(args):
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:89]:
```suggestion
        for idx, mod in enumerate(args):
```
we iterate modules here only?

File: kornia/augmentation/container/image.py [hunk 7]
```diff
@@ -116,17 +119,121 @@ def __init__(
...
+        self._validate_mix_augmentation(*args, validate_args=validate_args)
+        self.has_mix_augmentation = len(self.__get_mix_indices__(iter(args))) > 0
+
+    def update_attribute(
+        self,
+        same_on_batch: Optional[bool] = None,
+        return_transform: Optional[bool] = None,
+        keepdim: Optional[bool] = None,
+    ) -> None:
+        for arg in self.children():
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:131]:
```suggestion
        for mod in self.children():
```

Comment (edgarriba) [reply kornia/augmentation/container/image.py:131]: or `aug`

File: kornia/augmentation/container/image.py [hunk 8]
```diff
@@ -116,17 +119,121 @@ def __init__(
...
+                    arg.keepdim = keepdim
+
+    @property
+    def same_on_batch(self) -> Optional[bool]:
+        return self._same_on_batch
+
+    @same_on_batch.setter
+    def same_on_batch(self, same_on_batch: Optional[bool]) -> None:
+        self._same_on_batch = same_on_batch
+        self.update_attribute(same_on_batch=same_on_batch)
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:147]: are the others set `None` then ?

File: kornia/augmentation/container/image.py [hunk 9]
```diff
@@ -116,17 +119,121 @@ def __init__(
...
+        self._params = []
+
+    def _validate_mix_augmentation(self, *args: nn.Module, validate_args: bool = True):
+        mix_count = 0
+        for arg in args:
+            if isinstance(arg, (MixAugmentationBase,)):
+                mix_count += 1
+        if self.random_apply is False and mix_count > 1 and validate_args:
+            raise ValueError(
+                f"Multiple mix augmentation is prohibited without enabling random_apply. Detected {mix_count}."
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:177]:
```suggestion
                f"Multiple mix augmentation is supported without enabling ``random_apply``. Detected {mix_count}."
```

File: kornia/augmentation/container/image.py [hunk 10]
```diff
@@ -116,17 +119,121 @@ def __init__(
...
+            )
+        if mix_count > 1 and validate_args:
+            warnings.warn(
+                f"Multiple ({mix_count}) mix augmentation detected and at most one mix augmenation can"
+                "be applied at each forward. To silence this warning, please set `validate_args` to False.",
+                category=UserWarning,
+            )
+
+    def __get_mix_indices__(self, args: Iterator) -> List[int]:
```
Comment (edgarriba) [inline kornia/augmentation/container/image.py:186]: could you add at least one line function definition?

File: kornia/augmentation/container/patch.py [hunk 11]
```diff
@@ -219,93 +222,140 @@ def restore_from_patches(
...
+            [out_param.append(PatchParamItem(indices, p)) for p, _ in params]  # type: ignore
+            # "append" of "list" does not return a value
+        elif not self.same_on_batch:
+            params = self.get_parameter_sequence(batch_shape[0] * batch_shape[1])
+            [out_param.append(PatchParamItem([i], p)) for p, i in params]  # type: ignore
+            # "append" of "list" does not return a value
+        else:
+            params = self.get_parameter_sequence(batch_shape[1])
+            indices = torch.arange(0, batch_shape[0] * batch_shape[1], step=batch_shape[1])
+            [out_param.append(PatchParamItem((indices + i).numpy().tolist(), p)) for p, i in params]  # type: ignore
```
Comment (edgarriba) [inline kornia/augmentation/container/patch.py:239]:
```suggestion
            [out_param.append(PatchParamItem((indices + i).tolist(), p)) for p, i in params]  # type: ignore
```

File: kornia/augmentation/augmentation.py (hunk 2 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:1922]: I will update the typing. This was only meant to maintain the devices and dtypes across different tensors.

File: kornia/augmentation/container/image.py (hunk 3 above)
Comment (shijianjian) [reply kornia/augmentation/container/image.py:16]: When we only want to create the operation without specifying the parameter.

File: kornia/augmentation/container/image.py (hunk 5 above)
Comment (shijianjian) [reply kornia/augmentation/container/image.py:41]: The parenthesis meant optionally. LOL. I will try make it more clear.

File: kornia/augmentation/container/image.py (hunk 6 above)
Comment (shijianjian) [reply kornia/augmentation/container/image.py:89]: yes

File: kornia/augmentation/container/image.py (hunk 8 above)
Comment (shijianjian) [reply kornia/augmentation/container/image.py:147]: Nope. Nothing will happen if None is passes.

File: kornia/augmentation/container/image.py (hunk 10 above)
Comment (shijianjian) [reply kornia/augmentation/container/image.py:186]: I will refactor those mess soon. Pretty ugly now.

**Review by edgarriba — APPROVED** [review]
